File: pc/link_crypto.py
"""P-256 ECDH + HKDF-SHA256 + AES-GCM for post-HELLO USB link sealing."""
from __future__ import annotations

import hashlib
import struct

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from protocol import MAGIC

CURVE = ec.SECP256R1()
INFO = b"usb-bridge-link-v1"
GCM_TAG_LEN = 16


def generate_keypair():
    priv = ec.generate_private_key(CURVE)
    return priv, public_bytes(priv.public_key())


def public_bytes(pub) -> bytes:
    return pub.public_bytes(
        serialization.Encoding.DER,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )


def private_bytes(priv) -> bytes:
    return priv.private_bytes(
        serialization.Encoding.DER,
        serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption(),
    )


def load_private(der: bytes):
    return serialization.load_der_private_key(der, password=None)


def load_public(der: bytes):
    return serialization.load_der_public_key(der)


def pc_id_of(pubkey_der: bytes) -> str:
    return hashlib.sha256(pubkey_der).hexdigest()


def short_id(pc_id: str) -> str:
    return pc_id[:16] if pc_id else "?"


def derive_session_key(shared: bytes, token: str, pc_id: str) -> bytes:
    salt = bytes.fromhex(pc_id) if len(pc_id) == 64 else pc_id.encode("utf-8")
    return HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        info=INFO,
    ).derive(shared + token.encode("utf-8"))


def ecdh(priv, peer_pub_der: bytes) -> bytes:
    peer = load_public(peer_pub_der)
    shared = priv.exchange(ec.ECDH(), peer)
    # Left-pad / trim to 32 bytes so Android KeyAgreement and Python match
    if len(shared) >= 32:
        return shared[-32:]
    return shared.rjust(32, b"\x00")


class LinkSeal:
    """AES-GCM seal/unseal with per-direction counters (12-byte big-endian nonce)."""

    def __init__(self, key: bytes):
        if len(key) != 32:
            raise ValueError("session key must be 32 bytes")
        self._aes = AESGCM(key)
        self.send_ctr = 0
        self.recv_ctr = 0

    def seal_payload(self, msg_type: int, header_bytes: bytes, payload: bytes) -> bytes:
        inner = (struct.pack(">H", len(header_bytes)) + header_bytes
                 + struct.pack(">I", len(payload)) + payload)
        nonce = self.send_ctr.to_bytes(12, "big")
        self.send_ctr += 1
        aad = MAGIC + bytes([msg_type & 0xFF])
        return self._aes.encrypt(nonce, inner, aad)

    def unseal_payload(self, msg_type: int, ciphertext: bytes) -> tuple[bytes, bytes]:
        nonce = self.recv_ctr.to_bytes(12, "big")
        self.recv_ctr += 1
        aad = MAGIC + bytes([msg_type & 0xFF])
        inner = self._aes.decrypt(nonce, ciphertext, aad)
        if len(inner) < 6:
            raise IOError("sealed frame too short")
        hlen = struct.unpack(">H", inner[:2])[0]
        if len(inner) < 2 + hlen + 4:
            raise IOError("sealed frame truncated header")
        header_bytes = inner[2:2 + hlen]
        plen = struct.unpack(">I", inner[2 + hlen:6 + hlen])[0]
        if len(inner) < 6 + hlen + plen:
            raise IOError("sealed frame truncated payload")
        payload = inner[6 + hlen:6 + hlen + plen]
        return header_bytes, payload

File: android/app/src/main/java/com/example/usbbridge/LinkCrypto.kt
package com.example.usbbridge

import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** P-256 ECDH + HKDF-SHA256 + AES-GCM; mirrors pc/link_crypto.py. */
object LinkCrypto {
    private val MAGIC = byteArrayOf(0xAB.toByte(), 0xCD.toByte())
    private val INFO = "usb-bridge-link-v1".toByteArray(Charsets.UTF_8)

    fun generateKeyPair(): KeyPair {
        val kpg = KeyPairGenerator.getInstance("EC")
        kpg.initialize(ECGenParameterSpec("secp256r1"))
        return kpg.generateKeyPair()
    }

    fun publicDer(pub: PublicKey): ByteArray = pub.encoded

    fun loadPublic(der: ByteArray): PublicKey =
        KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(der))

    fun loadPrivate(der: ByteArray): PrivateKey =
        KeyFactory.getInstance("EC").generatePrivate(PKCS8EncodedKeySpec(der))

    fun pcIdOf(pubkeyDer: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(pubkeyDer)
            .joinToString("") { "%02x".format(it) }

    fun shortId(pcId: String): String =
        if (pcId.length >= 16) pcId.substring(0, 16) else pcId

    fun ecdh(priv: PrivateKey, peerPubDer: ByteArray): ByteArray {
        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(priv)
        ka.doPhase(loadPublic(peerPubDer), true)
        val shared = ka.generateSecret()
        // Left-pad / trim to 32 bytes so Python cryptography and Android match
        return when {
            shared.size >= 32 -> shared.copyOfRange(shared.size - 32, shared.size)
            else -> ByteArray(32 - shared.size) + shared
        }
    }

    fun deriveSessionKey(shared: ByteArray, token: String, pcId: String): ByteArray {
        val salt = if (pcId.length == 64) {
            pcId.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
        } else {
            pcId.toByteArray(Charsets.UTF_8)
        }
        val ikm = shared + token.toByteArray(Charsets.UTF_8)
        return hkdfSha256(ikm, salt, INFO, 32)
    }

    /** HKDF-Extract + Expand (RFC 5869) with SHA-256. */
    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        val actualSalt = if (salt.isEmpty()) ByteArray(32) else salt
        mac.init(SecretKeySpec(actualSalt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        val result = ByteArray(length)
        var t = ByteArray(0)
        var offset = 0
        var counter = 1
        while (offset < length) {
            mac.init(SecretKeySpec(prk, "HmacSHA256"))
            mac.update(t)
            mac.update(info)
            mac.update(counter.toByte())
            t = mac.doFinal()
            val copy = minOf(t.size, length - offset)
            System.arraycopy(t, 0, result, offset, copy)
            offset += copy
            counter++
        }
        return result
    }

    class Seal(key: ByteArray) {
        private val keySpec = SecretKeySpec(key, "AES")
        private var sendCtr = 0L
        private var recvCtr = 0L

        fun sealPayload(msgType: Int, headerBytes: ByteArray, payload: ByteArray): ByteArray {
            val inner = ByteArray(2 + headerBytes.size + 4 + payload.size)
            inner[0] = ((headerBytes.size shr 8) and 0xFF).toByte()
            inner[1] = (headerBytes.size and 0xFF).toByte()
            headerBytes.copyInto(inner, 2)
            val o = 2 + headerBytes.size
            inner[o] = ((payload.size shr 24) and 0xFF).toByte()
            inner[o + 1] = ((payload.size shr 16) and 0xFF).toByte()
            inner[o + 2] = ((payload.size shr 8) and 0xFF).toByte()
            inner[o + 3] = (payload.size and 0xFF).toByte()
            payload.copyInto(inner, o + 4)
            val nonce = counterNonce(sendCtr++)
            val aad = byteArrayOf(MAGIC[0], MAGIC[1], (msgType and 0xFF).toByte())
            val c = Cipher.getInstance("AES/GCM/NoPadding")
            c.init(Cipher.ENCRYPT_MODE, keySpec, GCMParameterSpec(128, nonce))
            c.updateAAD(aad)
            return c.doFinal(inner)
        }

        fun unsealPayload(msgType: Int, ciphertext: ByteArray): Pair<ByteArray, ByteArray> {
            val nonce = counterNonce(recvCtr++)
            val aad = byteArrayOf(MAGIC[0], MAGIC[1], (msgType and 0xFF).toByte())
            val c = Cipher.getInstance("AES/GCM/NoPadding")
            c.init(Cipher.DECRYPT_MODE, keySpec, GCMParameterSpec(128, nonce))
            c.updateAAD(aad)
            val inner = c.doFinal(ciphertext)
            require(inner.size >= 6) { "sealed frame too short" }
            val hlen = ((inner[0].toInt() and 0xFF) shl 8) or (inner[1].toInt() and 0xFF)
            require(inner.size >= 2 + hlen + 4) { "sealed frame truncated" }
            val headerBytes = inner.copyOfRange(2, 2 + hlen)
            val o = 2 + hlen
            val plen = ((inner[o].toInt() and 0xFF) shl 24) or
                    ((inner[o + 1].toInt() and 0xFF) shl 16) or
                    ((inner[o + 2].toInt() and 0xFF) shl 8) or
                    (inner[o + 3].toInt() and 0xFF)
            require(inner.size >= o + 4 + plen) { "sealed frame truncated payload" }
            val payload = inner.copyOfRange(o + 4, o + 4 + plen)
            return headerBytes to payload
        }

        private fun counterNonce(ctr: Long): ByteArray {
            val n = ByteArray(12)
            for (i in 0 until 8) {
                n[11 - i] = ((ctr shr (8 * i)) and 0xFF).toByte()
            }
            return n
        }
    }
}
